### ISSUE TITLE
ln -sfn is NOT atomic

### ISSUE THREAD

**Issue by ab** [mina-deploy/mina#16]
The "current" link will be briefly nonexistent during a deploy, which would break things like serving static files. Note the `unlink()` before `symlink()` in the strace below. For an atomic replacement, you can do `ln` and then `mv`, as in `ln -sfn 2 current.new && mv -fv current.new current`.

```
$ mkdir 1 2
$ ln -s 1 current
$ strace ln -sfn 2 current
execve("/bin/ln", ["ln", "-sfn", "2", "current"], [/* 52 vars */]) = 0
...
lstat("current", {st_mode=S_IFLNK|0777, st_size=1, ...}) = 0
stat("2", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
symlink("2", "current")                 = -1 EEXIST (File exists)
unlink("current")                       = 0
symlink("2", "current")                 = 0
close(0)                                = 0
close(1)                                = 0
close(2)                                = 0
exit_group(0)                           = ?
```

**Comment by apeschar** [comment]
For those arriving on this page from [Google](https://www.google.com/search?client=firefox-b-d&q=link+sfn+atomic), it's good to know that since [2018-10-28](https://fossies.org/linux/coreutils/ChangeLog) coreutils' `ln` actually does an atomic link-then-rename operation:

~~~
$ strace ln -sfn test current 2>&1 | grep -P 'test|current'
execve("/bin/ln", ["ln", "-sfn", "test", "current"], 0x7ffda9a9c0a8 /* 27 vars */) = 0
lstat("current", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
lstat("current", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
stat("test", 0x7ffca1f24cf0)            = -1 ENOENT (No such file or directory)
symlinkat("test", AT_FDCWD, "current")  = -1 EEXIST (File exists)
symlinkat("test", AT_FDCWD, "CutvsIlr") = 0
renameat(AT_FDCWD, "CutvsIlr", AT_FDCWD, "current") = 0
~~~

In other words, you can safely use `ln -sfn` and there's no need for `ln` followed by `mv`.

**Comment by JuliusJaenchen** [comment]
Thank you @apeschar!
